### PR TITLE
Aidl new method #6533

### PR DESCRIPTION
**Pull request by MadWasp79**
and documentation

### CODE REVIEW

**Comment by vshcherb** [comment]
1. RoutingHelper ???	public boolean removeListener(IRouteInformationListener lt){
2. Concurrent modification listeners / updateListeners
3.

File: OsmAnd/src/net/osmand/aidl/OsmandAidlService.java [hunk 1]
```diff
@@ -88,16 +89,31 @@
...
 	HandlerThread mHandlerThread = new HandlerThread("OsmAndAidlServiceThread");
 
-	private long updateCallbackId = 0;
+	private static long aidlCallbackId = 0;
 
 	private OsmandApplication getApp() {
 		return (OsmandApplication) getApplication();
 	}
 
+	public static void addAidlCallback(IOsmAndAidlCallback callback, long id, long key) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/aidl/OsmandAidlService.java]: here should be no id parameter, updateCallbackId

File: OsmAnd/src/net/osmand/aidl/OsmandAidlService.java [hunk 2]
```diff
@@ -88,16 +89,31 @@
...
-	private Map<Long, IOsmAndAidlCallback> callbacks;
+	public static final int KEY_ON_UPDATE = 1;
+	public static final int KEY_ON_NAV_DATA_UPDATE = 2;
+
+	private static Map<Long, AidlCallbackParams> callbacks;
 	private Handler mHandler = null;
 	HandlerThread mHandlerThread = new HandlerThread("OsmAndAidlServiceThread");
 
-	private long updateCallbackId = 0;
+	private static long aidlCallbackId = 0;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/aidl/OsmandAidlService.java]: no static

File: OsmAnd/src/net/osmand/aidl/OsmandAidlService.java [hunk 3]
```diff
@@ -88,16 +89,31 @@
 
 	private static final String DATA_KEY_RESULT_SET = "resultSet";
 
-	private Map<Long, IOsmAndAidlCallback> callbacks;
+	public static final int KEY_ON_UPDATE = 1;
+	public static final int KEY_ON_NAV_DATA_UPDATE = 2;
+
+	private static Map<Long, AidlCallbackParams> callbacks;
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/aidl/OsmandAidlService.java]: no static

File: OsmAnd/src/net/osmand/aidl/OsmandAidlService.java [hunk 4]
```diff
@@ -88,16 +89,31 @@
...
 
 	private OsmandApplication getApp() {
 		return (OsmandApplication) getApplication();
 	}
 
+	public static void addAidlCallback(IOsmAndAidlCallback callback, long id, long key) {
+		callbacks.put(id, new AidlCallbackParams(callback, key));
+	}
+
+	public static boolean removeAidlCallback (long id){
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/aidl/OsmandAidlService.java]: no static

File: OsmAnd/src/net/osmand/aidl/OsmandAidlService.java (hunk 3 above)
Comment (vshcherb) [reply OsmAnd/src/net/osmand/aidl/OsmandAidlService.java]: construct object here

File: OsmAnd/src/net/osmand/aidl/OsmandAidlApi.java [hunk 5]
```diff
@@ -1918,7 +1925,38 @@ boolean changePluginState(PluginParams params) {
...
+					directionInfo.setTurnType(TurnType.OFFR);
+					directionInfo.setDistanceTo((int) rh.getRouteDeviation());
+				} else {
+					NextDirectionInfo ndi = rh.getNextRouteDirectionInfo(baseNdi, true);
+					if (ndi != null && ndi.distanceTo > 0 && ndi.directionInfo != null) {
+						directionInfo.setDistanceTo(ndi.distanceTo);
+						directionInfo.setTurnType(ndi.directionInfo.getTurnType().getValue());
+					}
+				}
+				for (Entry<Long, OsmandAidlService.AidlCallbackParams> cb : OsmandAidlService.getAidlCallbacks().entrySet()) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/aidl/OsmandAidlApi.java]: valueset?

File: OsmAnd/src/net/osmand/aidl/OsmandAidlService.java (hunk 1 above)
Comment (MadWasp79) [reply OsmAnd/src/net/osmand/aidl/OsmandAidlService.java]: +

File: OsmAnd/src/net/osmand/aidl/OsmandAidlApi.java (hunk 5 above)
Comment (MadWasp79) [reply OsmAnd/src/net/osmand/aidl/OsmandAidlApi.java]: done

File: OsmAnd/src/net/osmand/aidl/OsmandAidlService.java (hunk 4 above)
Comment (MadWasp79) [reply OsmAnd/src/net/osmand/aidl/OsmandAidlService.java]: removed those methods and add listener. Setting service's listener to aidl api in onCreate

File: OsmAnd/src/net/osmand/aidl/OsmandAidlService.java (hunk 3 above)
Comment (MadWasp79) [reply OsmAnd/src/net/osmand/aidl/OsmandAidlService.java]: fixed

File: OsmAnd/src/net/osmand/aidl/OsmandAidlService.java (hunk 2 above)
Comment (MadWasp79) [reply OsmAnd/src/net/osmand/aidl/OsmandAidlService.java]: done

File: OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java [hunk 6]
```diff
@@ -502,6 +531,20 @@ private boolean updateCurrentRouteStatus(Location currentLocation, float posTole
 				route.updateCurrentRoute(newCurrentRoute + 1);
 				currentRoute = newCurrentRoute + 1;
 				app.getNotificationHelper().refreshNotification(NotificationType.NAVIGATION);
+				if (!updateListeners.isEmpty()) {
+					ArrayList<WeakReference<IRoutingDataUpdateListener>> tmp = new ArrayList<>(updateListeners);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java]: Here no need to modify list! Could be very slow!

File: OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java [hunk 7]
```diff
@@ -265,6 +267,33 @@ public void addListener(IRouteInformationListener l){
 		transportRoutingHelper.addListener(l);
 	}
 
+	public void addDataUpdateListener(IRoutingDataUpdateListener listener) {
+		List<WeakReference<IRoutingDataUpdateListener>> copyList = new ArrayList<>(updateListeners);
+		Iterator<WeakReference<IRoutingDataUpdateListener>> it = copyList.iterator();
+		while (it.hasNext()) {
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java]: Looks very similar to removeDataUpdateListener, 5-7 lines of code could be moved to a common private method

File: OsmAnd/src/net/osmand/aidl/AidlCallbackListener.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package net.osmand.aidl;
+
+import java.util.Map;
+
+public interface AidlCallbackListener{
+	long addAidlCallback(IOsmAndAidlCallback callback, int key);
```
Comment (vshcherb) [inline OsmAnd/src/net/osmand/aidl/AidlCallbackListener.java]: Interface needs documentation and proper indents between methods

File: OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java (hunk 6 above)
Comment (MadWasp79) [reply OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java]: +

File: OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java (hunk 7 above)
Comment (MadWasp79) [reply OsmAnd/src/net/osmand/plus/routing/RoutingHelper.java]: +

File: OsmAnd/src/net/osmand/aidl/AidlCallbackListener.java (hunk 8 above)
Comment (MadWasp79) [reply OsmAnd/src/net/osmand/aidl/AidlCallbackListener.java]: fixed & added

**Pull request closed without merging**